# _Input/Output_

En este capítulo, estudiaremos a fondo el manejo de _strings_, _bytes_, arreglos de _bytes_, archivos, y _context managers_.

## _Strings_

Hasta ahora, hemos trabajado muchas veces con _strings_ que, como sabemos, corresponden a una secuencia inmutable de caracteres. En Python 3, todos los _strings_ se representan en [Unicode](https://es.wikipedia.org/wiki/Unicode), codificación que permite representar virtualmente cualquier caracter en cualquier lenguaje. Luego, veremos más detalles sobre Unicode. Entonces, pensemos que en Python un _string_ es una secuencia inmutable de caracteres Unicode. A continuación, algunas formas distintas de crear un _string_ en Python:

In [1]:
a = "programando"
b = 'mucho'
c = '''un string
con múltiples
lineas'''

d = """Multiples con
     comillas dobles"""
e = ("Tres" " strings" " juntos")
f = "un string " + "concatenado"

print(a)
print(b)
print(c)
print(d)
print(e)
print(f)

programando
mucho
un string
con múltiples
lineas
Multiples con
     comillas dobles
Tres strings juntos
un string concatenado


La clase `str` tiene muchos métodos para manipular _strings_. Aquí, con la función [`dir`](https://docs.python.org/3/library/functions.html#dir), podemos obtener la lista:

In [2]:
print(dir(str))

['__add__', '__class__', '__contains__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__getnewargs__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__mod__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__rmod__', '__rmul__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', 'capitalize', 'casefold', 'center', 'count', 'encode', 'endswith', 'expandtabs', 'find', 'format', 'format_map', 'index', 'isalnum', 'isalpha', 'isdecimal', 'isdigit', 'isidentifier', 'islower', 'isnumeric', 'isprintable', 'isspace', 'istitle', 'isupper', 'join', 'ljust', 'lower', 'lstrip', 'maketrans', 'partition', 'replace', 'rfind', 'rindex', 'rjust', 'rpartition', 'rsplit', 'rstrip', 'split', 'splitlines', 'startswith', 'strip', 'swapcase', 'title', 'translate', 'upper', 'zfill']


Ahora, algunos ejemplos concretos:

In [3]:
# El método isalpha retorna True si todos los caracteres del string
# están en el alfabeto de algún lenguaje.
print("abñ".isalpha()) 

# Si hay algún número, espacio o puntuación dentro del string, retornará False.
print("t/".isalpha())

# El método isdigit retorna True si todos los caracteres en el string
# son dígitos numéricos.
print("34".isdigit())

s = "estoy programando"
print(s.startswith("est"))
print(s.endswith("do"))

# Devuelve el índice donde comienza en s la secuencia que se pasa como argumento.
print(s.find("y p"))

# El método index retorna el índice donde comienza la secuencia.
# Acepta dos argumentos opcionales:
# - la posición inicial donde comenzar la búsqueda
# - y la posición final (hasta dónde llega buscando)
print(s.index("y", 4, 10))
print(s.index("p", 5, 10))

True
False
True
True
True
4
4
6


Otros métodos que actúan sobre _strings_ retornan un _string_ nuevo (sin modificar el original), ya que recordemos que los _strings_ son objetos **inmutables**. Veamos algunos ejemplos.

In [4]:
s = "hola a todos, cómo están"
s2 = s.split(' ')
print(s2)
s3 = '#'.join(s2)
print(s3)
print(s.replace(' ', '**'))
print(s)

['hola', 'a', 'todos,', 'cómo', 'están']
hola#a#todos,#cómo#están
hola**a**todos,**cómo**están
hola a todos, cómo están


Como ya hemos visto muchas veces, podemos insertar valores de variables dentro de un _string_ usando `format`.

In [5]:
nombre = 'Juan Pérez'
nota = 4.5
if nota >= 4.0:
    resultado = 'aprobado'
else:
    resultado = 'reprobado'

template = "Hola {0}, estás {1}. Tu nota fue un {2}."
template.format(nombre, resultado, nota)

'Hola Juan Pérez, estás aprobado. Tu nota fue un 4.5.'

Podemos lograr el mismo resultado también utilizando _f-strings_, los cuales hemos utilizado varias veces a lo largo del curso. 

Los _f-strings_ son una forma de formatear _strings_ que fue añadida en Python 3.6 y permite añadir expresiones de Python directamente en un _string_. Para indicarle a Python que utilizaremos esta funcionalida, debemos antemponer el string con la letra ``f`` (de ahí el nombre _f-string_)

In [6]:
f"Hola {nombre}, estás {resultado}. Tu nota fue un {nota}."

'Hola Juan Pérez, estás aprobado. Tu nota fue un 4.5.'

Si queremos incluir las _llaves_ (`{`, `}`) dentro del _string_, podemos agregar un _escape character_, que permite invocar una interpretación alternativa de los caracteres siguientes. Más concretamente, las llaves en este caso, se utilizan para encapsular la variable que queremos imprimir. Sin embargo, la representación alternativa, sería la representación literal de la llave. Luego, para este caso, si queremos imprimir una llave, esto se logrará con una doble llave.

Veamos un ejemplo. Digamos que buscamos imprimir una simple definición de una clase en Java.

In [7]:
template = """
public class {0} 
{{
       public static void main(String[] args) 
       {{
           System.out.println({1});
       }} 
}}"""

print(template.format("MiClase", "'hola mundo'"));


public class MiClase 
{
       public static void main(String[] args) 
       {
           System.out.println('hola mundo');
       } 
}


A veces queremos incluir muchas variables dentro de un _string_, esto hace que sea difícil recordar el orden en que debemos escribirlas dentro de la función `format`. Una solución es usar argumentos con _keywords_ en la función `format`:

In [8]:
print("{} {label} {}".format("x", "y", label="z"))

x z y


In [9]:
template = """
From: <{from_email}>
To: <{to_email}>
Subject: {subject}
{message}
"""

print(template.format(
    from_email = "kpb@ing.puc.cl",
    to_email = "cualquiera@example.com",
    message = "\nEste es un mail de prueba.\n\n"
              "Espero que el mensaje te sea de mucha utilidad!", # ver como Python concatena esto automáticamente 
    subject = "Este correo es urgente")
    )


From: <kpb@ing.puc.cl>
To: <cualquiera@example.com>
Subject: Este correo es urgente

Este es un mail de prueba.

Espero que el mensaje te sea de mucha utilidad!



Podemos, incluso, usar contenedores como listas, tuplas o diccionarios como argumentos dentro de la función `format`.

In [10]:
emails = ("a@ejemplo.com", "b@ejemplo.com")
message = {'subject': "Tienes un correo", 'message': "Este es un correo para ti"}
template = """
From: <{0[0]}>
To: <{0[1]}>
Subject: {message[subject]} {message[message]}
""" 
print(template.format(emails, message=message))


From: <a@ejemplo.com>
To: <b@ejemplo.com>
Subject: Tienes un correo Este es un correo para ti



Además, podemos usar un diccionario con listas e indexar la lista dentro del _string_.

In [11]:
mensaje = {
    "emails": ["yo@ejemplo.com", "tu@ejemplo.com"],
    "subject": "mira este correo",
    "message": "Sorry no era tan importante"
}

template = """
From: <{0[emails][0]}>
To: <{0[emails][1]}>
Subject: {0[subject]}
{0[message]}"""

print(template.format(mensaje))


From: <yo@ejemplo.com>
To: <tu@ejemplo.com>
Subject: mira este correo
Sorry no era tan importante


Lo mismo se puede replicar con _f-strings_, sin embargo tenemos que utilizar las _keys_ como fueron definidas originalmente, es decir como _strings_.

In [12]:
print(
f"""From: <{mensaje['emails'][0]}>
To: <{mensaje['emails'][1]}>
Subject: {mensaje['subject']}
{mensaje['message']}"""
)

From: <yo@ejemplo.com>
To: <tu@ejemplo.com>
Subject: mira este correo
Sorry no era tan importante


Esto puede ser aún mejor: podemos pasar cualquier objeto como argumento, por ejemplo, una instancia de una clase. Luego, dentro del _string_ podemos acceder a cualquiera de los atributos del objeto.

In [13]:
class EMail:
    def __init__(self, from_addr, to_addr, subject, message):
        self.from_addr = from_addr
        self.to_addr = to_addr
        self.subject = subject
        self.message = message
        
email = EMail("a@ejemplo.com", "b@ejemplo.com","Tienes un correo","\nEl mensaje es inútil\n\nSaludos")
template = """
From: <{0.from_addr}>
To: <{0.to_addr}>
Subject: {0.subject}
{0.message}"""

print(template.format(email))


From: <a@ejemplo.com>
To: <b@ejemplo.com>
Subject: Tienes un correo

El mensaje es inútil

Saludos


Y lo mismo también puede ser replicado con _f-strings_.

In [14]:
print(f"""
From: <{email.from_addr}>
To: <{email.to_addr}>
Subject: {email.subject}
{email.message}""")


From: <a@ejemplo.com>
To: <b@ejemplo.com>
Subject: Tienes un correo

El mensaje es inútil

Saludos


También podemos mejorar el formato de los _strings_ que se imprimen. Por ejemplo, en casos como la impresión de una tabla con datos, muchas veces queremos que datos pertenecientes a la misma variable se vean alineados en columnas:

In [15]:
compra = [('leche', 2, 120), ('pan', 3.5, 800), ('arroz', 1.75, 960)]

print("PRODUCTO  CANTIDAD   PRECIO   SUBTOTAL")
for producto, precio, cantidad in compra:
    subtotal = precio * cantidad
    print("{0:8s}{1: ^9d}    ${2: <8.2f}${3: >7.2f}".format(producto, cantidad, precio, subtotal))

PRODUCTO  CANTIDAD   PRECIO   SUBTOTAL
leche      120       $2.00    $ 240.00
pan        800       $3.50    $2800.00
arroz      960       $1.75    $1680.00


Podemos lograr lo mismo utilizando _f-strings_.

In [16]:
print("PRODUCTO  CANTIDAD   PRECIO   SUBTOTAL")
for producto, precio, cantidad in compra:
    subtotal = precio * cantidad
    print(f"{producto:8s}{cantidad: ^9d}    ${precio: <8.2f}${subtotal: >7.2f}")

PRODUCTO  CANTIDAD   PRECIO   SUBTOTAL
leche      120       $2.00    $ 240.00
pan        800       $3.50    $2800.00
arroz      960       $1.75    $1680.00


Notar que, dentro de cada llave, existe un item tipo diccionario; es decir, antes de los dos puntos va el índice del argumento dentro de la función `format`. Después de los dos puntos, por ejemplo, `8s`, significa que el dato es un _string_ de ocho caracteres. Por defecto, si el _string_ es más corto que los ocho caracteres, el resto se llenará con espacios (por la derecha). No olvidar que también, por defecto, si el _string_ que ingresamos es más largo que los 8 caracteres, este no será truncado:

In [17]:
compra = [('lecheeeeeeeeeeeeeee', 2, 120), ('pan', 3.5, 800), ('arroz', 1.75, 960)]

print("PRODUCTO  CANTIDAD   PRECIO   SUBTOTAL")
for producto, precio, cantidad in compra:
    subtotal = precio * cantidad
    print("{0:8s}{1: ^9d}    ${2: <8.2f}${3: >7.2f}".format(producto, cantidad, precio, subtotal))

PRODUCTO  CANTIDAD   PRECIO   SUBTOTAL
lecheeeeeeeeeeeeeee   120       $2.00    $ 240.00
pan        800       $3.50    $2800.00
arroz      960       $1.75    $1680.00


Podemos cambiar esta situación obligando a que el _string_ sea truncado si se pasa del largo máximo. Basta con agregar un punto y la precisión luego del número que indica el largo del _string_. La precisión para tipos no numéricos indica el largo máximo de caracteres usados para este campo.

In [16]:
compra = [('lecheeeeeeeeeeeeeee', 2, 120), ('pan', 3.5, 800), ('arroz', 1.75, 960)]

print("PRODUCTO  CANTIDAD   PRECIO   SUBTOTAL")
for producto, precio, cantidad in compra:
    subtotal = precio * cantidad
    print("{0:i<8.8s}{1:a^9d}    ${2: <8.2f}${3: >7.2f}".format(producto, cantidad, precio, subtotal))

PRODUCTO  CANTIDAD   PRECIO   SUBTOTAL
lecheeeeaaa120aaa    $2.00    $ 240.00
paniiiiiaaa800aaa    $3.50    $2800.00
arroziiiaaa960aaa    $1.75    $1680.00


Para la cantidad de producto el formato es `{1:^d}`, el 1 corresponde al índice del argumento en la función `format`, el espacio después de los dos puntos dice que los lugares vacíos deben ser llenados con espacios (en los tipos enteros, por defecto, se llenará con ceros), el símbolo `^` es para que el número quede centrado en el espacio disponible, `9d` significa que será un entero de hasta nueve dígitos. Notar que siempre el orden de estos parámetros (aunque son opcionales) debe ser de izquierda a derecha después de los dos puntos: primero el caracter para llenar los espacios vacíos, después el alineamiento, después el tamaño y finalmente el tipo.

Para el precio por ejemplo, `{2: <8.2f}` significa que el dato se leerá del tercer argumento de la función `format`, luego los lugares que queden libres se llenarán con espacios, el símbolo `<` significa que el alineamiento es a la izquierda, el formato será un _float_ de hasta ocho caracteres, con dos decimales.
  
De la misma forma, para el subtotal, `{3: >7.2f}` significa que el dato se sacará del cuarto argumento dentro de la función `format`, el caracter de llenado será espacio, el alineamiento es a la derecha, será un _float_ de siete dígitos, dos de ellos decimales e incluyendo el `.` como carácter.

## _Bytes_ y I/O

Al comienzo del capítulo, dijimos que los _strings_ en Python eran una colección de caracteres Unicode inmutables. Unicode no es realmente un formato válido de almacenamiento de datos: muchas veces leemos información correspondiente a algún _string_ desde un archivo o un [_socket_](https://es.wikipedia.org/wiki/Socket_de_Internet) en bytes, no en Unicode. Los _bytes_ son el formato de almacenamiento de más bajo nivel, representan una secuencia de 8 _bits_, descritas en general como un entero entre 0 y 255, un [hexadecimal](https://es.wikipedia.org/wiki/Sistema_hexadecimal) equivalente entre `0` y `FF`, o un literal (sólo se permiten los caracteres [ASCII](https://es.wikipedia.org/wiki/ASCII) para representar _bytes_). Los _bytes_ a secas pueden representar cualquier entidad, desde caracteres codificados de un _string_, o pixeles de una imagen. En general, necesitamos saber la forma en que fueron codificados para poder interpretar el tipo de datos correcto representados por los _bytes_. Por ejemplo, un patrón binario de 8 _bits_ (1 _byte_) puede corresponder a un carácter en particular si lo decodificamos como un ASCII, pero puede corresponder a un caracter completamente distinto si lo decodificamos como un caracter Unicode. 
 
En Python los _bytes_ se representan con el objeto tipo `bytes`. Para declarar que un objeto es un _byte_ simplemente se pone al comienzo del objeto una `b`. Por ejemplo:

In [17]:
caracteres = b'\x63\x6c\x69\x63\x68\xe9' # estamos diciendo que lo que está entre las comillas es un objeto de bytes
print(caracteres)
print(caracteres.decode("latin-1"))
caracteres = b"ab"
caracteres = b"\x61\x62"  # 61 y 62 es la representación en hexadecimal de los caracteres a y b, respectivamente
print(caracteres.decode("ascii"))
caracteres = bytes((97, 98))  # 97 y 98 corresponden al código ASCII de los caracteres a y b, respectivamente
print(caracteres)

b'clich\xe9'
cliché
ab
b'ab'


In [20]:
caracteres = b"áb" # esto genera un error ya que sólo se pueden usar literales ASCII para la creación de bytes

SyntaxError: bytes can only contain ASCII literal characters. (<ipython-input-20-f7f127161793>, line 1)

El símbolo escape `\x` indica que los siguientes dos caracteres después de la `x` corresponden a un _byte_ usando dígitos hexadecimales. Los bytes que coinciden con los bytes de ASCII son reconocidos inmediatamente, así cuando los tratamos de imprimir aparecen correctamente (_e.g._ `cliché`), el resto se imprime como hexadecimal. La `b` en la impresión nos recuerda que lo que está a la derecha es un objeto de bytes, no un _string_. La sentencia `caracteres.decode("latin-1")` decodifica la secuencia de  bytes usando el alfabeto `latin-1`.

El método `decode` retorna un _string_ normal (Unicode). Si, por ejemplo, hubiésemos usado otro alfabeto, habríamos obtenido otro _string_.

In [21]:
caracteres = b'\x63\x6c\x69\x63\x68\xe9'
print(caracteres.decode("latin-1"))
print(caracteres.decode("iso8859-5"))

cliché
clichщ


Para codificar un _string_ en distintos alfabetos, simplemente usamos el método `encode` de la clase `str`. Obviamente es necesario ingresar como argumento el conjunto de caracteres o alfabeto con que se quiere codificar.

In [18]:
characters = "estación"
print(characters.encode("UTF-8"))  # 8-bit Unicode Transformation Format
print(characters.encode("latin-1"))
print(characters.encode("CP437"))
print(characters.encode("latin-1"))

# No se puede codificar en ASCII el caracter "ó" ya que no existe dentro 
# de los 128 caracteres de ASCII
print(characters.encode("ascii"))  

b'estaci\xc3\xb3n'
b'estaci\xf3n'
b'estaci\xa2n'
b'estaci\xf3n'


UnicodeEncodeError: 'ascii' codec can't encode character '\xf3' in position 6: ordinal not in range(128)

El método `encode` nos ofrece opciones de cómo manejar el caso en que el _string_ que se quiere codificar no puede ser codificado con el alfabeto requerido. Estas opciones se ingresan a través del argumento opcional `errors`, donde los valores posibles son: `strict` (el valor por defecto), `replace`, `ignore` o `xmlcharrefreplace`. 

In [23]:
print(characters.encode("ascii", errors='replace'))  # en ascii se reemplaza el caracter desconocido con "?"
print(characters.encode("ascii", errors='ignore'))
print(characters.encode("ascii", errors='xmlcharrefreplace'))  # se crea una entidad xml que representa el caracter Unicode

b'estaci?n'
b'estacin'
b'estaci&#243;n'


En general, si queremos codificar un _string_ y no sabemos con qué alfabeto deberíamos codificar, lo mejor es usar UTF-8, ya que es _backwards_ compatible con ASCII: los primeros 128 caracteres de UTF-8 son los mismos que en ASCII. Recordar siempre que los objetos tipo _byte_ son **inmutables**.

## `bytearrays`

Tal como el nombre lo sugiere, los _bytearrays_ son arreglos de _bytes_, y a diferencia de los _bytes_ **son mutables**. Los _bytearrays_ se comportan como las listas: podemos indexar con la notación de _slices_, y también podemos ir agregando _bytes_ con el método `extend`. Para construir un _bytearray_ podemos ingresar un _byte_ inicial.

In [24]:
ba_1 = bytearray(b"holamundo")
print(ba_1)
print(ba_1[3:7])
ba_1[4:6] = b"\x15\xa3"
print(ba_1)
ba_1.extend(b"programa")
print(ba_1)
print(ba_1[0])  # Notar que aquí se imprime un entero, el ascii que corresponde a la letra "h"
print(bin(ba_1[0]))
print(bin(ba_1[0])[2:].zfill(8))

bytearray(b'holamundo')
bytearray(b'amun')
bytearray(b'hola\x15\xa3ndo')
bytearray(b'hola\x15\xa3ndoprograma')
104
0b1101000
01101000


Notar que la última línea es para imprimir directamente los _bits_ correspondientes al primer _byte_ (representado en el literal `h` o el entero 104). El `[2:]` es para partir desde la tercera posición, ya que las primeras dos posiciones contienen los caracteres `0b`, que simplemente indica que el formato es en binario (línea anterior). Al agregar `.zfill(8)` indicamos que se usarán 8 _bits_ para representar el _byte_, lo cual tiene sentido cuando hay ceros por el lado izquierdo y el _default_ no los muestra (línea anterior tiene sólo 7 bits después del `0b`). 

Un caracter de un _byte_ puede ser convertido a un entero usando la función `ord`.

In [25]:
print(ord(b"a"))
b = bytearray(b'abcdef')
b[3] = ord(b'g')  # La letra g tiene como código ascii el 103
b[4] = 68  # La letra D tiene como código ascii el 68, esto sería lo mismo que ingresar b[4] = ord(b'D')
print(b)

97
bytearray(b'abcgDf')


## I/O de archivos

Hasta ahora hemos operado con la lectura y escritura de archivos de texto; sin embargo, los sistemas operativos representan los archivos como secuencias de _bytes_, no como texto. Dado que leer _bytes_ y convertirlos a texto es una operación muy común en archivos, Python se encarga de manejar los _bytes_ que vienen o van transformándolos a la respectiva representación en _string_ con los _encoders_/_decoders_ correspondientes. La ya conocida función `open` nos permite además de abrir archivos, ingresar como argumentos el set de caracteres que se usará para codificar los _bytes_ y la estrategia que se debe seguir cuando aparezcan _bytes_ inconsistentes con el formato:

In [26]:
file = open('data/archivo_ejemplo', "r", encoding='ascii', errors='replace')
print(file.read())
file.close()

sorry pero ahora yo soy lo que habr�� dentro del archivo
yo me agregar�� al final



Veamos cómo cambia esto, si es que eligimos otro _encoding_.

In [27]:
file = open('data/archivo_ejemplo', "r", encoding='utf-8', errors='replace')
print(file.read())
file.close()

sorry pero ahora yo soy lo que habrá dentro del archivo
yo me agregaré al final



Ahora, escribiremos en un archivo el mismo contenido.

In [28]:
contenido = "sorry pero ahora yo soy lo que habrá dentro del archivo"
file = open("data/archivo_ejemplo_1", "w", encoding="utf-8", errors="replace")
file.write(contenido)
file.close()

Podemos también agregar contenido al final del archivo, reemplazando el modo de apertura del archivo, cambiando la `w` por una `a`, de _append_.

In [29]:
contenido = "\nyo me agregaré al final"
file = open("data/archivo_ejemplo_1", "a", encoding="utf-8", errors="replace")
file.write(contenido)
file.close()

file = open('data/archivo_ejemplo_1', "r", encoding='utf-8', errors='replace')
print(file.read())
file.close()

sorry pero ahora yo soy lo que habrá dentro del archivo
yo me agregaré al final


Para abrir un archivo como binario, simplemente debemos agregar una `b` por el lado derecho del modo de apertura. Por ejemplo, `wb` o `rb`. El archivo se comportará igual que un archivo de texto, sólo que sin la codificación automática de _byte_ a texto.

In [30]:
contenido = b"abcde12"
file = open("data/archivo_ejemplo_2", "wb")
file.write(contenido)
file.close()

file = open('data/archivo_ejemplo_2', "rb")
print(file.read())
file.close()

b'abcde12'


Podemos además concatenar _bytes_ simplemente con el operador suma. En el siguiente ejemplo, construimos un contenido dinámico para ser escrito en un archivo de _bytes_. Después leemos una cantidad fija de _bytes_ desde el mismo archivo:

In [31]:
num_lineas = 100

file = open("data/archivo_ejemplo_3", "wb")
for i in range(num_lineas):
    # A la función "bytes" debemos pasarle un iterable con el contenido a convertir 
    # por eso le pasamos el entero dentro de una lista  
    contenido = b"linea_" + bytes([i]) + b" abcde12 "                                             
    file.write(contenido)
file.close()

file = open('data/archivo_ejemplo_3', "rb")
# El número dentro de la función read nos dice el número de bytes que se van a leer del archivo
print(file.read(41))
file.close()

b'linea_\x00 abcde12 linea_\x01 abcde12 linea_\x02 a'


## _Context Manager_

Dado que siempre necesitamos cerrar un archivo después de usarlo, debemos considerar la posibilidad de que ocurran excepciones mientras el archivo está abierto. Una forma clara de hacerlo es cerrar el archivo dentro de la sentencia `finally` después de un `try`. El problema es que esto genera mucho código extra. Afortunadamente, en Python existe una forma de hacer lo mismo con menos código, a través de un _context manager_, que se encarga de ejecutar las sentencias `try` y `finally` sin la necesidad de llamarlas directamente. Sólo necesitamos llamar al archivo que abriremos con la sentencia `with`. Veamos un ejemplo.

In [32]:
with open("data/archivo_ejemplo_4", "r") as file:
    contenido = file.read()

El código anterior sería equivalente a hacer lo siguiente:

In [33]:
file = open("data/archivo_ejemplo_4", "r")
try:
    contenido = file.read()
finally:
    file.close()

Si ejecutamos `dir` en un objeto de tipo archivo:

In [34]:
file = open("data/archivo_ejemplo_4", "w")
print(dir(file))
file.close()

['_CHUNK_SIZE', '__class__', '__del__', '__delattr__', '__dict__', '__dir__', '__doc__', '__enter__', '__eq__', '__exit__', '__format__', '__ge__', '__getattribute__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__lt__', '__ne__', '__new__', '__next__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '_checkClosed', '_checkReadable', '_checkSeekable', '_checkWritable', '_finalizing', 'buffer', 'close', 'closed', 'detach', 'encoding', 'errors', 'fileno', 'flush', 'isatty', 'line_buffering', 'mode', 'name', 'newlines', 'read', 'readable', 'readline', 'readlines', 'seek', 'seekable', 'tell', 'truncate', 'writable', 'write', 'writelines']


Vemos que existen dos métodos llamados `__enter__` y `__exit__`. Estos dos métodos transforman el archivo en un _context manager_. El método `__exit__` asegura que el archivo será cerrado incluso si aparece una excepción mientras esté abierto. El método `__enter__` inicializa el archivo o realiza cualquier acción necesaria para ajustar el contexto del objeto.

Para asegurarnos que un archivo usará los métodos `__enter__` y `__exit__`, simplemente debemos llamar a la apertura del archivo con el método `with`.
  
Podemos crear nuestros propios _context managers_: simplemente creamos cualquier clase, agregamos los métodos `__enter__` y `__exit__` y podemos llamar a nuestra clase a través del método `with`. Del siguiente ejemplo, se puede ver cómo el método `__exit__` se ejecuta una vez que nos salimos del _scope_ de la sentencia `with`.

In [35]:
import string, random

class StringUpper(list): 
        
    def __enter__(self):
        return self
    
    def __exit__(self, exc_type, exc_value, traceback):
        for index, char in enumerate(self):
            self[index] = char.upper()

with StringUpper() as s_upper:
    for i in range(20):
        # Aquí seleccionamos, en forma aleatoria, un ascii en minúsculas
        # y las agregamos a la lista.
        s_upper.append(random.choice(string.ascii_lowercase))
    print(s_upper)
        
print(s_upper)

['r', 'p', 'p', 'y', 'x', 'd', 'g', 'w', 'p', 'e', 'p', 'n', 'p', 'f', 'g', 'f', 'i', 'y', 'f', 'j']
['R', 'P', 'P', 'Y', 'X', 'D', 'G', 'W', 'P', 'E', 'P', 'N', 'P', 'F', 'G', 'F', 'I', 'Y', 'F', 'J']


El código anterior simplemente corresponde a una clase que hereda de la clase `list`. Al implementar los métodos `__enter__` y `__exit__`, podemos instanciar la clase a través de un _context manager_. En este ejemplo en particular, el _context manager_ se encarga de transformar todos los caracteres ASCII de la lista a mayúsculas.

## Cómo emular archivos para I/O

Muchas veces tenemos que interactuar con algunos módulos de software que sólo leen y escriben sus datos desde y hacia archivos. Si queremos comunicar nuestro código que genera, por ejemplo, _strings_, para evitar tener que escribir nuestros datos en un archivo para que el otro programa los lea, podemos _emular_ el tener un archivo usando los módulos de Python, `StringIO` o `BytesIO`. El siguiente ejemplo muestra cómo usar estos módulos:

In [36]:
from io import StringIO, BytesIO
file_in = StringIO("información como texto y más") # aquí simulamos tener un archivo que contiene el string dado 
file_out = BytesIO() # aquí simulamos un archivo de Bytes para escribir la información

char = file_in.read(1)
while char:
    file_out.write(char.encode("ascii", "ignore"))
    char = file_in.read(1)

buffer_ = file_out.getvalue()
buffer_

b'informacin como texto y ms'

## Más información

Para conocer más sobre _bytes_, puedes visitar [este enlace](http://www.dummies.com/programming/electronics/digital-electronics-binary-basics/).